llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1915,6 +1967,147 @@ PackOp::reifyResultShapes(OpBuilder &builder,
+        int64_t tile = *maybeConstantTileSize;
+        int64_t sizeDim = inputType.getDimSize(dim) * tile;
+        inferredShape.push_back(sizeDim);
+      }
+    } else {
+      inferredShape.push_back(inputType.getShape()[dim]);
+    }
+  }
+  assert(inferredShape.size() == getOutputRank() &&
+         "expect inferredShape to match output rank");
+  return TypeSwitch<Type, ShapedType>(inputType)
+      .Case<RankedTensorType>([&](RankedTensorType t) -> ShapedType {
+        return RankedTensorType::get(inferredShape, inputType.getElementType());
+      })
+      .Case<MemRefType>([&](MemRefType t) -> ShapedType {
+        return MemRefType::get(inferredShape, inputType.getElementType());
+      })
+      .Default([&](Type t) {
+        llvm_unreachable("unexpected type");
+        return nullptr;
+      });
+}
+
+LogicalResult UnPackOp::verify() {
+  Operation *op = getOperation();
+  size_t numberOfBlockingFactors = getMixedTiles().size();
+  SmallVector<int64_t> dimsPos = extractFromI64ArrayAttr(getDimsPos());
+  if (failed(commonVerifierPackAndUnPackOp(*this))) {
+    return failure();
+  }
+  // TODO: implement padding semantics. Bail out for now.
+  if (getPaddingValue())
+    return op->emitError("padding semantics not implemented");
+
+  // Blocking factors must be less or equal than the input rank, and must
+  // match the number of `dims_pos`.
+  if (numberOfBlockingFactors > getOutputRank()) {
+    return op->emitError(
+        "blocking factors must be less or equal than the output rank");
+  }
+  // Require `dim_pos` to be in-bound. `dim_pos` carries the index of the
+  // dimensions to block.
+  if (!isInBound(dimsPos, getOutputRank()))
+    return op->emitError("out-of-bound position");
+  // Require input rank to match output rank + number of blocking factors.
+  if ((getOutputRank() + numberOfBlockingFactors) != getInputRank()) {
+    return op->emitError(
+        "input rank must equal output rank + blocking factors");
+  }
+  // Bail out if the tile does not divide the dimension fully. In the case of
+  // dynamic tile factors or dimensions, having a partial tile is undefined
+  // behavior. We will relax this constraint when we introduce padding
+  // semantics.
+  if (areNotFullTiles(getOutputShape(), getDimAndTileMapping())) {
+    return op->emitError(
+        "invalid tile factor provided. Only full tiles are supported");
+  }
+  // Verify result type against inferred type.
+  ShapedType expectedType = inferResultType();
+  if (expectedType != getOutputType()) {
+    return op->emitError(
+               "inferred type do not match provied output type. Expected ")
+           << expectedType << " but got: " << getOutputType();
+  }
+  return success();
+}
+
+// Implement the tiling interface. The number of loops equals the rank of the
+// input. All the loops are parallel.
+SmallVector<utils::IteratorType> UnPackOp::getLoopIteratorTypes() {
+  SmallVector<utils::IteratorType> iteratorTypes(getInputRank(),
```

Comment:
I think the loop iterator types should be same as the output rank. The inner tile dimensions are not part of the loop iterator types since these dimensions are not tiled. So similar to the `pack` operation, the unpack operations scalar implementation needs to generate explicit loops for "unpacking" the tile.